### PR TITLE
Create field boundary conditions using information from grid topology

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR aims to create a single `FieldBoundaryConditions` function to replace `HorizontallyPeriodicBCs` and `ChannelBCs`. This paves the way for boundary conditions to become a `Field` property (#606).

It's a pretty big breaking change so would be good to agree on the design and API before refactoring.

Side note: Ideally we would set `default_bc(::Flat) = nothing` but until we elide operations and halo filling along `Flat` dimensions, I think we should use `default_bc(::Flat) = PeriodicBC()`.

@glwagner Let me know if this looks like what we discussed. If so, I'll go ahead and change all the tests and examples to use it instead of `HorizontallyPeriodicBCs` and `ChannelBCs`.

Reminder to self: Documentation, especially model setup, will have to be revised as part of this PR.

### CODE REVIEW

File: src/BoundaryConditions/field_boundary_conditions.jl [hunk 1]
```diff
@@ -14,47 +14,50 @@ Construct a `FieldBoundaryConditions` using a `CoordinateBoundaryCondition` for
...
-    z = CoordinateBoundaryConditions(bottom, top)
+default_bc(::Periodic) = PeriodicBC()
+default_bc(::Bounded)  = NoFluxBC()
+
+valid_directions = (:west, :east, :south, :north, :bottom, :top)
+
+function FieldBoundaryConditions(grid; kwargs...)
+    kws = keys(kwargs)
+    for kw in kws
+        if kw ∉ valid_directions
```
Comment (glwagner) [inline src/BoundaryConditions/field_boundary_conditions.jl]: Another strategy would be to set all defaults in the constructor to 'nothing', and then reset them to default if they are nothing. The error the user receives would then bean ordinary KeywordError. This also may simplify subsequent validation.

File: src/BoundaryConditions/field_boundary_conditions.jl [hunk 2]
```diff
@@ -14,47 +14,50 @@ Construct a `FieldBoundaryConditions` using a `CoordinateBoundaryCondition` for
...
+        end
+    end
+
+    TX, TY, TZ = topology(grid)
+
+    if (TX isa Periodic || TX isa Flat) && (:west in kws || :east in kws)
+        e = "Cannot specify west or east boundary conditions with $TX topology in x-direction."
+        throw(ArgumentError(e))
+    else
+        west_bc = :west in kws ? kwargs[:west] : default_bc(TX)
```
Comment (glwagner) [inline src/BoundaryConditions/field_boundary_conditions.jl]: could let default_bc take two arguments: The topology and the current bc. Then define a fallback that throws an error, whitelist the bcs that are valid, and define a default for ::Nothing. Might be a bit simpler and more modular, less repeated code.

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/BoundaryConditions/field_boundary_conditions.jl]: I ended up assigning default boundary conditions in the constructor which cleaned up the function quite a bit.

**Comment by ali-ramadhan** [comment]
I've refactored a bit more so that we can use `FieldBoundaryConditions` to construct velocity, tracer, diffusivity, pressure, and tendency boundary conditions. This is done by specifying the `field_type` required kwarg when calling `FieldBoundaryConditions`.

```julia
u_bcs = FieldBoundaryConditions(grid, field_type=:velocity)
T_bcs = FieldBoundaryConditions(grid, field_type=:tracer, top=FluxBoundaryCondition(0.1))
```

An alternative design:
```julia
VelocityFieldBoundaryConditions(grid; kwargs...) =
    FieldBoundaryConditions(grid, field_type=:velocity, kwargs...)
TracerFieldBoundaryConditions(grid; kwargs...) =
    FieldBoundaryConditions(grid, field_type=:tracer, kwargs...)
```

This will allow us to properly set default velocity and tracer boundary conditions with a single `FieldBoundaryConditions` constructor. It will also allow us to clean up the code a bit and remove some other functions like `PressureBoundaryConditions` and `DiffusivitiesBoundaryConditions` from `solution_and_model_boundary_conditions.jl`.

@glwagner Let me know what you think.

**Comment by ali-ramadhan** [comment]
I added tests and updated the documentation so will merge once tests pass.

I also added more readable (i.e. reads like how you would say it) boundary condition constructors, e.g.
```julia
fbc = FluxBoundaryCondition(-1e-3)
vbc = ValueBoundaryCondition(25)
```
